cpp/src/arrow/compute/codegen_internal.h

```diff
@@ -0,0 +1,31 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/api_vector.h"
+#include "arrow/result.h"
+#include "arrow/type_fwd.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+ARROW_EXPORT Result<TypeHolder> FirstType(KernelContext*,
```

Comment:
Why is this function moved here but not the other similar ones in `kernels/codegen_internal.h`?

Reply:
this was the only function required for the core functionality kernels and was the only one strictly required. I agree that the rest would probably make sense together so I've moved all of them from `arrow/compute/kernels/codegen_internal` to `arrow/compute/codegen_internal` and have added the necessary `ARROW_EXPORT`s